contracts/contracts/vault/OETHVaultCore.sol

```diff
@@ -125,24 +137,338 @@ contract OETHVaultCore is VaultCore {
+     * If one of the requests is not claimable, the whole transaction will revert with `Queue pending liquidity`.
+     * If one of the requests is not older than 10 minutes,
+     * the whole transaction will revert with `Claim delay not met`.
+     * @param _requestIds Unique ID of each withdrawal request
+     * @return amounts Amount of WETH received for each request
+     * @return totalAmount Total amount of WETH transferred to the withdrawer
+     */
+    function claimWithdrawals(uint256[] memory _requestIds)
+        external
+        whenNotCapitalPaused
+        nonReentrant
+        returns (uint256[] memory amounts, uint256 totalAmount)
+    {
+        // Just call the Dripper instead of looping through _requestIds to find the highest id
+        // and checking it's queued amount is > the queue's claimable amount.
+
+        // Stream any harvested rewards (WETH) that are available to the Vault
+        IDripper(dripper).collect();
+
+        // Add any WETH from the Dripper to the withdrawal queue
+        _addWithdrawalQueueLiquidity();
+
+        amounts = new uint256[](_requestIds.length);
+        for (uint256 i = 0; i < _requestIds.length; ++i) {
+            amounts[i] = _claimWithdrawal(_requestIds[i]);
+            totalAmount += amounts[i];
+        }
+
+        // transfer all the claimed WETH from the vault to the withdrawer
+        IERC20(weth).safeTransfer(msg.sender, totalAmount);
+
+        // Prevent insolvency
+        _postRedeem(totalAmount);
+    }
+
+    function _claimWithdrawal(uint256 requestId)
+        internal
+        returns (uint256 amount)
+    {
+        // Load the structs from storage into memory
+        WithdrawalRequest memory request = withdrawalRequests[requestId];
+        WithdrawalQueueMetadata memory queue = withdrawalQueueMetadata;
+
+        require(
+            request.timestamp + CLAIM_DELAY <= block.timestamp,
+            "Claim delay not met"
+        );
+        // If there isn't enough reserved liquidity in the queue to claim
+        require(request.queued <= queue.claimable, "Queue pending liquidity");
+        require(request.withdrawer == msg.sender, "Not requester");
+        require(request.claimed == false, "Already claimed");
+
+        // Store the request as claimed
+        withdrawalRequests[requestId].claimed = true;
+        // Store the updated claimed amount
+        withdrawalQueueMetadata.claimed = queue.claimed + request.amount;
+
+        emit WithdrawalClaimed(msg.sender, requestId, request.amount);
+
+        return request.amount;
+    }
+
+    /// @notice Collects harvested rewards from the Dripper as WETH then
+    /// adds WETH to the withdrawal queue if there is a funding shortfall.
+    /// @dev is called from the Native Staking strategy when validator withdrawals are processed.
+    /// It also called before any WETH is allocated to a strategy.
+    function addWithdrawalQueueLiquidity() external {
```

Comment:
Let's add `nonReentrant` to this one, since it's updating state.

Reply:
Done

Reply:
Hmm, this is not as simple as just adding `nonReentrant` to `OETHVaultCore.addWithdrawalQueueLiquidity` as `OETHVaultAdmin` does external calls to `OETHVaultCore.addWithdrawalQueueLiquidity` which will fail the reentry. These external calls are in `swapCollateral`, `withdrawFromStrategy`, `withdrawAllFromStrategy` and `withdrawAllFromStrategies`

Reply:
So our options are:
1. No reentry protection on `addWithdrawalQueueLiquidity`
2. duplicate the internal `_addWithdrawalQueueLiquidity` function on `OETHVaultAdmin`
3. change `nonReentrant` to ignore reentry from itself. ie the proxy contract
4. move `_addWithdrawalQueueLiquidity` up to the common `VaultStorage` contract
4. Add a new contract for the common functions across `OETHVaultCore` and `OETHVaultAdmin`. ie `_addWithdrawalQueueLiquidity` and `_wethAvailable`

Reply:
Here's the what the inheritance of the OETH Vault currently looks like

![OETHVaultHierarchy](https://github.com/user-attachments/assets/03047830-9f2d-4fb8-8deb-c6c67f008874)

Reply:
Adding a `OETHVaultCommon` contract with `_addWithdrawalQueueLiquidity` and `_wethAvailable` seems ok. This is what the inheritance looks like 

![OETHVaultHierarchy](https://github.com/user-attachments/assets/d01216ee-f1b1-42f1-865c-04a449b12df6)

I've done the changes in a separate PR if we want to do it https://github.com/OriginProtocol/origin-dollar/pull/2153

Reply:
What if we just did something like `require(msg.sender == addres(this))` on this method. There shouldn't be a reason someone outside the protocol actually needs to call this externally.

Reply:
All the vault actions that add WETH call `_addWithdrawalQueueLiquidity`. But we do have the Native Staking Strategy that will transfer WETH to the vault when it accounts for validator withdrawals. The strategy does not call `addWithdrawalQueueLiquidity` on the vault. A separate tx in the Defender Action has to call `addWithdrawalQueueLiquidity`.

Reply:
@DanielVF any preference over which option we take to resolve this one?

Reply:
I would lean then towards leaving it as it then, since we have to have this called externally by both the vault and by other code. Not worth adding complexity in the service of extra defensiveness.

Reply:
The other option would be to have two external functions - one nonRentrant, and callable by the world, and the other vault only without the reentrancy check.

Reply:
How about we use a library with the two duplicate functions? I can do it on a PR so you can see what it looks like

Reply:
Here's the PR with the common internal functions in a library https://github.com/OriginProtocol/origin-dollar/pull/2161